0-module-and-module-source.emu

```diff
@@ -0,0 +1,614 @@
+                1. Assert: Evaluate has already been invoked on _moduleRecord_ and successfully completed.
+                1. Let _namespace_ be Completion(GetModuleNamespace(_moduleRecord_)).
+                1. If _namespace_ is an abrupt completion, then
+                  1. Perform ! Call(_promiseCapability_.[[Reject]], *undefined*, &laquo; _namespace_.[[Value]] &raquo;).
+                1. Else,
+                  1. Perform ! Call(_promiseCapability_.[[Resolve]], *undefined*, &laquo; _namespace_.[[Value]] &raquo;).
+                1. Return ~unused~.
+              1. Let _onFulfilled_ be CreateBuiltinFunction(_fulfilledClosure_, 0, *""*, &laquo; &raquo;).
+              1. Let _rejectedClosure_ be a new Abstract Closure with parameters (_error_) that captures _promiseCapability_ and performs the following steps when called:
+                1. Perform ! Call(_promiseCapability_.[[Reject]], *undefined*, &laquo; _error_ &raquo;).
+                1. Return ~unused~.
+              1. Let _onRejected_ be CreateBuiltinFunction(_rejectedClosure_, 0, *""*, &laquo; &raquo;).
+              1. Perform PerformPromiseThen(_innerPromise_, _onFulfilled_, _onRejected_).
+              1. Return ~unused~.
+            </emu-alg>
+
+            <emu-note type="editor">
+                This Abstract Operation is analogous to FinishDynamicImport. A future refactor can use this abstract operation as the foundation for other existings usages of FinishDynamicImport.
+            </emu-note>
+          </emu-clause>
+
+
+        <emu-clause id="sec-createcustomimportmeta" type="abstract operation">
+            <h1>
+                CreateCustomImportMeta (
+                    _module_: A Module Record,
+                    _meta_: An object,
+                ): ~unused~
+            </h1>
+            <dl class="header">
+              <dt>description</dt>
+              <dd>From a given _importMeta_, this abstract operation creates a List of Records for the qualifying values that must be present when a module body accesses import.meta syntax.</dd>
+            </dl>
+            <emu-alg>
+                1. Assert: _meta_ is an Object.
+                1. Assert: _module_ is a Module Record.
+                1. Assert: _module_.[[ImportMeta]] is ~empty~.
+                1. Let _importMeta_ to OrdinaryObjectCreate(*null*).
+                1. Let _keys_ be ? <emu-meta effects="user-code">_from_.[[OwnPropertyKeys]]()</emu-meta>.
+                1. For each element _nextKey_ of _keys_, do
+                  1. Let _desc_ be ? <emu-meta effects="user-code">_from_.[[GetOwnProperty]]</emu-meta>(_nextKey_).
+                  1. If _desc_ is not *undefined* and _desc_.[[Enumerable]] is *true*, then
+                    1. Let _nextValue_ be ? Get(_from_, _nextKey_).
+                    1. Perform ! CreateDataPropertyOrThrow(_importMeta_, _nextKey_,_nextValue_).
+                1. Perform HostFinalizeImportMeta(_importMeta_, _module_).
```

Comment:
Why it will call the host hook? This will prevent virtualization if the host did something observable.

Reply:
`HostFinalizeImportMeta`: https://tc39.es/ecma262/#sec-hostfinalizeimportmeta

> Most [hosts](https://tc39.es/ecma262/#host) will be able to simply define [HostGetImportMetaProperties](https://tc39.es/ecma262/#sec-hostgetimportmetaproperties), and leave HostFinalizeImportMeta with its default behaviour. However, HostFinalizeImportMeta provides an "escape hatch" for [hosts](https://tc39.es/ecma262/#host) which need to directly manipulate the object before it is exposed to ECMAScript code.

Basically, the creation of the meta object involves two steps, one of them needs virtualization (`HostGetImportMetaProperties`), the other one `HostFinalizeImportMeta` doesn't.

Reply:
No, since it does not call the `HostGetImportMetaProperties`, it should not call  `HostFinalizeImportMeta` either.